### ISSUE TITLE
Cluster mode and distribution of event processing

### ISSUE THREAD

**Issue by mvoros** [cap-js-community/event-queue#259]
Hi,

I am trying to maximize the throughput of my app. On high level, the app provides API that receives a list of transactions. The transactions are dumped into a DB table and then an event is raised that processes these events asynchronously. Redis is used as a backend for queues. I was doing load testing on my laptop and really quickly realized that I am hitting node's limitation of having a single process.

cds serve now supports switch `--workers` that uses node's cluster mode to create multiple processes and then distributes incoming requests across these processes. This seems to be working fine for API calls. I can see that the incoming calls are being distributed across multiple workers and I can see quicker processing of API calls. 

Unfortunatly, this does not seem to help with event processing. All events are still being processed by single process only. I can see this by logging process ID after my events get processed. The process ID changes between event batches but I don't see parallel processing of the events.  

Is there a mechanism how to use multiple processes for event processing in parallel? In general, how can one achieve the highest throughput for event processing? What should the set up look like when we want to have multiple parallel event processors? 

Thanks

**Comment by soccermax** [comment]
Hi @mvoros,

yes the locking mechanism ensures by design that only one node process is processing one exact event combination (type + subtype). Up until now the `parallelEventProcessing` parameter was enough to archive parallelization in the same process. Processing the same event combination in different node processes is currently not possible.
What do you mean with `hitting node's limitation of having a single process`? CPU load, memory, blocked event-loop, IO?

Enabling processing of the same event across node processes would be a bigger change in the architecture but we could talk that through if you're interested in a discussion.

What you could do right now is: Add more event configurations for the same event-processor (impl). The API receives a list of transactions distributes the incoming load to the different event configurations. Would that be a feasable option for you?

Best regards,
Max

**Comment by mvoros** [comment]
Are you saying that even with single process, there should be concurrent processing of events by single node? I am not sure that I see it or it's not enough for my case. The events are basically doing some simple calculation and then updating Redis cache. I don't expect that Redis is my bottleneck. It seems to me that basically having one single process with one event loop for processing events is not enough. But I have not done any proper analysis.

If I understand it correctly, even when if I deploy the app to multiple nodes (e.g. Cloud Foundry on BTP), only single node will be processing these events. Is this correct? Ideally, multiple processes could be processing the same event and the lock would be used only for fetching the events from the queue. 

I expect this burst behaviour for my app where I get lots of transactions and then I want to process them as quickly as possible. It does not seem that I can achive this out of the box. 

Regading your workaround. I had this idea as well. I may try to implement it to see if my throughput improves. I can randomly assign events to the "queues". Is it possible to do this in runtime? I can see in the documentation an example like this `const eventConfig = config.getEventConfig("HealthCheck", "DB");`. Ideally here, I would clone my template event definition during startup and create versions of the event dynamically.

Thanks

**Comment by soccermax** [comment]
> Are you saying that even with single process, there should be concurrent processing of events by single node?

No there should not. The current locking mechanism does not allow this.

> If I understand it correctly, even when if I deploy the app to multiple nodes (e.g. Cloud Foundry on BTP), only single node will be processing these events. Is this correct? Ideally, multiple processes could be processing the same event and the lock would be used only for fetching the events from the queue.

Currently not no. But this could be implemented. I could draw up a concept and we could discuss it.

> Regading your workaround. I had this idea as well. I may try to implement it to see if my throughput improves. I can randomly assign events to the "queues". Is it possible to do this in runtime? I can see in the documentation an example like this const eventConfig = config.getEventConfig("HealthCheck", "DB");. Ideally here, I would clone my template event definition during startup and create versions of the event dynamically.


We could try this first and see if this works. If this increases the throughput as you expect, I could implement a configuration to allow processing of the same event on different node processes/ cloud foundry instances. 
For the workaround: I would propose to do the clone directly in the eventConfig.yml because every running instance need be aware of the multiple event configuration otherwise instance 0 might know 2 configurations and instance 1 only one.
So you could do something like:
```yml
events:
  - type: PROCESSING
    subType: A
    impl: ./srv/EventProcessor.js
    load: 1
    parallelEventProcessing: 10
    checkForNextChunk: true

  - type: PROCESSING
    subType: B
    impl: ./srv/EventProcessor.js
    load: 1
    parallelEventProcessing: 10
    checkForNextChunk: true

  - type: PROCESSING
    subType: C
    impl: ./srv/EventProcessor.js
    load: 1
    parallelEventProcessing: 10
    checkForNextChunk: true

  - type: PROCESSING
    subType: D
    impl: ./srv/EventProcessor.js
    load: 1
    parallelEventProcessing: 10
    checkForNextChunk: true
```

And your handle could distribute them:
```js
const eventConfigs = config.events.filter(event => event.type === "PROCESSING");
// here you know how many configurations you have and you could randomly use one of the configurations
```

Does this make sense to you? If not we could also schedule a short meeting to discuss the options.

**Comment by mvoros** [comment]
Ok. This makes sense. I will try to run some experiments but not sure when I get time to do it. Thanks

**Comment by soccermax** [comment]
One more thing that comes to my mind: are you testing this with sqlite/better sqlite?
Because this could also lower throughput because sqlite is using only one transaction which can be the bottleneck in your situtation.

**Comment by mvoros** [comment]
Thanks for the tip. This was the first bottleneck that I found. I am testing against PostgreSQL. The throughput is much higher than with SQLite. But it does not seem that DB is bottleneck anymore.

**Comment by mvoros** [comment]
I did the initial round of testing with multiple queues. I defined my event 4 times and added random distribution to these events when submitting the events for processing (simple assignment with uniform distribution using Math.random). Afterwards I ran cds serve with addition `--workers`. This spawns 4 node processes as my testing VM has access to 4 virtual processors. The result of the test is positive as it seems to be scaling linearly with 4 processes. I could see all 4 events being processes in parallel. Niether DB nor Redis seemed to be a bottleneck yet.    

What I don't like about the solution is that the events have to be statically defined. One possible solution could be that this distribution happens transparently via event configuration and publishEvent API. Alternatively, it would be nice to have an API that can be used to define events dynamically at the start of the system. 

In general, it would be nice to have a solution that allows parallel processing of events of the same type when there are multiple node instances or node cluster mode is used. 

Thanks for the help

**Comment by soccermax** [comment]
> What I don't like about the solution is that the events have to be statically defined.

I completely agree that manual distribution via different event configurations is not a real solution. This was just for testing if the general idea works.

> Alternatively, it would be nice to have an API that can be used to define events dynamically at the start of the system.

I thought about this already a couple of times that it would make sense to have a API to dynamically create event configurations. But I personally think for the problem you describe in this issue this is not the best solution. 

> In general, it would be nice to have a solution that allows parallel processing of events of the same type when there are multiple node instances or node cluster mode is used.

Yes I would also go for this solution. The idea that I have in my mind is to provide a new flag for the event configuration:
```
events:
  - type: Mail
    subType: Single
    impl: ./srv/EventQueueMail
    load: 10
    parallelEventProcessing: 5
    processOnAllInstances: true // new flag
```
The flag `processOnAllInstances` would be an optional opt-in possibility to process the same event on all available instances. What do you think? I'll look into this.

**Comment by soccermax** [comment]
For your testing with `--workers` do you know if there is any kind of instance number available? In Cloud foundry this is possible but for your local testing we would need sth similar with the CDS `--workers` approach.

**Comment by mvoros** [comment]
I think it would be better to have an integer instead of a boolean parameter. The default value would be 1 as it's now. When higher number is provided, the events would be split into corresponding queues (event0, event1, event2, ...). So even if there is not enough processes, it does not cause any issues as eventually all events will be processed. There should be APIs to adjust this dynamically during runtime. This would also allow to do some rudimentary load balancing. 

 Regarding API for event configuration. I think it would be really great to have it. You can imagine scenario where you want to process events as quickly as possible but you want to partition events so you have some kind of ordering per partition. E.g. you receive data per location/customer/vendor and you want to have single queue per these entities. This is not doable right now but would be relatively easy with API to define queues.

I will have to check how to get number of instances in cluster mode. This may not be possible as the individual nodes are runninng without any knowledge of other processes.

**Comment by soccermax** [comment]
Going with an integer is also a nice option. With that there are a few more questions to tackle:
- if you give it the number 3 and you have only 2 instances would you expect that:
   - option 1: 
     - instance 0 is processing two queues in parallel
     - instance 1 is processing one queue in parallel
   - option 2:
     - instance 0 is processing one queue in parallel
     - instance 1 is processing one queue in parallel (queue 3 needs to wait until one is finished)

Creation of API configuration is a bigger topic where quite a few things need to be tackled. Let me outline some interesting questions, feel free to share your thoughts on that:
- if you create a dynamic event configuration during runtime what happens if your app restarts?
  - event config yaml is not single source of truth anymore 
  - Sure based on the database I still know the dynamic event types but the configuration is missing
    - I could add a new column and persist the configuration for dynamic event types in the event queue table
    - I could add another table for the configuration but this would create new problems
      - how to get rid of orphaned event configurations or how to change event configurations?
      - From the top of my head I'm not a fan of this


Just wanted to dump my initial thoughts on this; I'll keep thinking about this and also discuss it with other architects. Feel free to add your thoughts and ideas as well

**Comment by mvoros** [comment]
Regarding the first topic. Ideally, the consumers would consume the events in a round robin fashion to avoid long delays for some of the events. Maybe splitting events into separete queues with indices is not a good idea. Maybe the right approach is that multiple consumers can consume events from a single queue. I am not sure how this can be achieved as I did not study the internals yet but there are various queues based on top of Postgres that can achieve this using SKIP LOCKED. I am not sure if this is supported by HANA though. 

Regarding the second topic. I can think of two solutions. One would be a callback to fill in event configuration when it's not found in the current state. So as a developer, I could reconstruct the config. Alternatively, with these partitions, I would still statically define a template event and then programatically create versions of it based on the needs (e.g. one version per location). Hence only the parent relationship could be stored in the event and look up for config would traverse up the chain until a valid configuration is found.

**Comment by soccermax** [comment]
I discussed this with some architects in our area, and we came up with the following solutions/ideas:
It would make sense (especially for the single-tenant case) to allow disabling exclusive locking of events. We currently do not see the need for having this integer-based because this can also be controlled with the load parameter. In your scenario, where you have only one event combination, you could set the load to 100, and with that, all instances would work in parallel on the events of the same type. Would that work for your use case?

Regarding the dynamic event definition API: this is a topic I’m going to work on in the future, but we would like to collect some more use cases to better shape the idea and how it should be built. For the use cases you describe: such things are already possible today. In your event processor, you can redefine the method: [clusterQueueEntries](https://cap-js-community.github.io/event-queue/implement-event/#clusterqueueentries). With this method, you can cluster events by all kinds of properties, such as customer/vendor/location. Every event supports a referenceEntityKey, which can be used for properties like customer. Of course, you do not have all the features, like defining the load and other properties, but I think such configurations are rather static and can therefore be defined in the config.yaml during development. As soon as I start with the implementation of the API, I’ll get back to you to get your feedback. Thanks for your input on this topic.

**Comment by mvoros** [comment]
I am not sure I understand the first part of the comment. So you want to allow disabling exclusive locking of events. Hence multiple proceses could grab the events of the same type. Only thing that I don't understand is if this will still prevent two processes processing the same event instance. If yes, it should work fine. I am happy to test this if this is available in some form.   

Thanks

**Comment by soccermax** [comment]
Yes it's always ensured that the processing of the same event instance is never going to happen. This is ensured by `RECORD TABLE LOCKS`. The skip exclusive lock means that there can be more processors (instances or workers) for the same event type.

**Comment by soccermax** [comment]
@mvoros would mind giving the branch of the linked PR a try in your local setup and see if everything works and how the performance is?
You can add the following in your package json as event-queue dependency:
```
"@cap-js-community/event-queue": "cap-js-community/event-queue#fb/skipExclusiveLocking",
```


If everything works as expected I would release a new version before Christmas :)

```yaml
  - type: SingleTenant
    subType: SkipExclusiveLocking
    impl: ./test/asset/EventQueueTest
    load: 1
    selectMaxChunkSize: 1
    skipExclusiveLocking: true // new flag to allow concurrent processing
    checkForNextChunk: true // BTW: with the next bigger release I'll switch the default to true for this flag
```

**Comment by mvoros** [comment]
Hi,

sorry for late response. Business and sickness did not allow me to find some time. But now I did the following tests. 

In all test cases, I am running CAP app inside Linux VM. VM is also running PostgreSQL and Redis in separate docker containers. I am importing 2M transactions that get dumped into DB and then processed via events. The transactions are submitted in chunks of 500 transactions and therfore 4k API calls are performed. These calls result in 4k events. The events perform some basic calculation and then update Redis cache. The CAP app is executed with option --workers. This triggers 4 different node processes as VM has 4 virtual CPUs. 

Test #1: manually define queue 0-3 and split via random assignment generated events. The event definition has the following parameters: `load: 1`, `checkForNextChunk: true`. I could validate that once API calls are finished, all 4 processes are being used to process events and total runtime was around 750 seconds. The longest waiting time for event to be processed since submission was 522 seconds. 

Test #2: simple event definition using the provided branch. The event definition has the following parameters: `load: 100`, `checkForNextChunk: true`, `skipExclusiveLocking: true`. I could validate that once API calls are finished, all 4 processes are being used to process events and total runtime was around 1025 seconds. The longest waiting time for event to be processed since submission was 807 seconds.

So it seems to work fine without any code adjustments, but the throughtput went down and total run time increased bu one third. My theory would be that this is cauesed by the `selectMaxChunkSize: 1`. Is this option required when processing in paralle? I think batching events for processing would help. It's also not clear to me what load exactly means. 

Thanks

**Comment by soccermax** [comment]
Hi @mvoros,

thanks for the tests. I'm sorry for the confusion with `selectMaxChunkSize: 1 ` this was just my unit test configuration. You can define the parameter as you want, there is no restriction. So I would assume that with the bigger chunk size we get the same result or even slighlty better results as with your **Test 1**.
Can you confirm that? If this works for you I would release the new version containing that flag

**Comment by mvoros** [comment]
OK. I re-ran my test without specifying `selectMaxChunkSize: 1`. It is basically comparable to a custom solution. It was 3% slower but I executed the test runs only once and there may be some other noise. So it looks OK from my side.

I also did run my integration tests and nothing got broken. 

Thanks

**Comment by soccermax** [comment]
I renamed the flag to `multiInstanceProcessing`. In my opinion, this name is more understandable.

Regarding your question about `load`:

> It's also not clear to me what "load" exactly means.

I need to polish the documentation a bit more on that. I described it [here](https://cap-js-community.github.io/event-queue/load-balancing/), but I admit it’s too abstract. I’ll improve it, but essentially, it works as follows:

- Every application instance can absorb a load of 100.  
- Each event configuration has a configurable `load` parameter (the one we're discussing).  
- Imagine you have three events:  
  - The first with a load of 50.  
  - The second with a load of 40.  
  - The third with a load of 20.  

This means that events 1 and 3 can be processed in parallel on one instance. However, event 2 cannot run in parallel with both 1 and 3—it would need to wait until either 1 or 3 finishes or another application instance with available capacity takes over.  

In your case, this isn’t relevant, but in multi-tenancy mode, the `load` would multiply with the number of tenants. Does this make sense to you?

As for the flag: When `multiInstanceProcessing` is enabled, events can run in parallel. I recommend setting the `load` to at least 51. Otherwise, the same event configuration could run twice on the same instance. Since the maximum load is 100, an event with a `load` of 50 could fit twice into the same instance.

**Comment by mvoros** [comment]
OK. It's clearer now. But why woul;d be a problem if same instance schdules same event multiple times? Is this just not recommended but it would still work?

**Comment by soccermax** [comment]
This works without any problems.
It was just a recommendation for your use case. Within the same instance, you can control concurrency using the event parameter: `parallelEventProcessing`. With this, you should be able to achieve maximized resource utilization (e.g., CPU) and throughput. However, you can use it either way; it was just a recommendation.

**Comment by soccermax** [comment]
release is published: https://github.com/cap-js-community/event-queue/releases/tag/v1.8.1
